Assets/HoloToolkit/Input/Scripts/Cursor/Cursor.cs

```diff
@@ -0,0 +1,478 @@
+        /// </summary>
+        [Header("Tranform References")]
+        public Transform PrimaryCursorVisual;
+
+        public Vector3 Position
+        {
+            get { return transform.position; }
+        }
+        
+        public Quaternion Rotation
+        {
+            get { return transform.rotation; }
+        }
+
+        public Vector3 LocalScale
+        {
+            get { return transform.localScale; }
+        }
+
+        /// <summary>
+        /// Indicates if hand is current in the view
+        /// </summary>
+        protected bool IsHandVisible;
+
+        /// <summary>
+        /// Indicates air tap down
+        /// </summary>
+        protected bool IsInputSourceDown;
+
+        protected GameObject TargetedObject;
+        protected ICursorModifier TargetedCursorModifier;
+
+        private bool isRegisteredToGazeManager = false;
+        private bool isInputRegistered = false;
+
+        private uint visibleHandsCount = 0;
+        private bool isVisible = true;
+
+        private GazeManager gazeManager;
+
+        /// <summary>
+        /// Position, scale and rotational goals for cursor
+        /// </summary>
+        private Vector3 targetPosition;
+        private Vector3 targetScale;
+        private Quaternion targetRotation;
+
+        /// <summary>
+        /// Indicates if the cursor should be visible
+        /// </summary>
+        public bool IsVisible
+        {
+            set
+            {
+                isVisible = value;
+                SetVisiblity(isVisible);
+            }
+        }
+
+#region MonoBehaviour Functions
+
+        private void Awake()
+        {
+            // Use the setter to update visibility of the cursor at startup based on user preferences
+            IsVisible = isVisible;
+            SetVisiblity(isVisible);
+        }
+
+        private void Start()
+        {
+            gazeManager = GazeManager.Instance;
+
+            RegisterGazeManager();
+            RegisterInput();
+        }
+
+        private void Update()
+        {
+            UpdateCursorState();
+            UpdateCursorTransform();
+        }
+
+        /// <summary>
+        /// Override for enable functions
+        /// </summary>
+        protected virtual void OnEnable(){}
+
+        /// <summary>
+        /// Override for disable functions
+        /// </summary>
+        protected virtual void OnDisable()
+        {
+            TargetedObject = null;
+            TargetedCursorModifier = null;
+            visibleHandsCount = 0;
+        }
+
+        private void OnDestroy()
+        {
+            UnregisterInput();
+            UnregisterGazeManager();
+        }
+
+#endregion
+
+        /// <summary>
+        /// Register to events from the gaze manager, if not already registered.
+        /// </summary>
+        private void RegisterGazeManager()
+        {
+            if (!isRegisteredToGazeManager && gazeManager != null)
```

Comment:
Agreed, the state management here isn't needed anymore. I removed it, and merged registration / unregistration in two protected virtual methods called RegisterManagers and UnregisterManagers